### PR TITLE
Benchmarking Provider tests 

### PR DESCRIPTION
**Pull request by mjeffryes**
## Minimum viable benchmarking for providers.

This introduces a test framework for recording benchmarks by collecting trace spans from ProgramTest. After collecting spans, we filter to spans that are most relevant to the provider itself and record into a benchmark file format in order to support comparing the results with [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat)

### example benchstat of master vs v5 release candidate

```
                                                                                       │    v4.txt    │                v5.txt                     │
                                                                                       │    sec/op    │    sec/op           vs base               │
/dns-recordset/preview-initial/ResourceProvider/GetPluginInfo/                           574.2µ ± 10%    602.5µ ±  2%  +4.93% (p=0.001 n=180+120)
/dns-recordset/preview-initial/newPlugin/gcp                                             209.7m ±  0%    210.5m ±  1%  +0.37% (p=0.007 n=60+40)
/dns-recordset/preview-initial/ResourceProvider/CheckConfig/pulumi:providers:gcp         249.6m ±  2%    251.8m ±  3%       ~ (p=0.135 n=60+40)
/dns-recordset/preview-initial/ResourceProvider/Configure/                               161.5m ±  2%    164.6m ±  2%       ~ (p=0.171 n=60+40)
/dns-recordset/preview-initial/ResourceProvider/Check/gcp:dns/managedZone:ManagedZone    702.2µ ± 19%    791.7µ ± 10%       ~ (p=0.056 n=30+20)
/dns-recordset/preview-initial/ResourceProvider/Create/gcp:dns/managedZone:ManagedZone   766.7µ ± 19%    867.6µ ± 13%       ~ (p=0.151 n=30+20)
/dns-recordset/preview-initial/ResourceProvider/Check/gcp:compute/instance:Instance      928.3µ ± 13%   1016.6µ ± 10%  +9.51% (p=0.049 n=30+20)
/dns-recordset/preview-initial/ResourceProvider/Create/gcp:compute/instance:Instance     3.475m ± 19%    3.592m ±  7%       ~ (p=0.064 n=30+20)
/dns-recordset/preview-initial/ResourceProvider/Check/gcp:dns/recordSet:RecordSet        1.438m ± 19%    1.352m ± 29%       ~ (p=0.976 n=30+20)
/dns-recordset/preview-initial/ResourceProvider/Create/gcp:dns/recordSet:RecordSet       1.056m ± 10%    1.075m ± 27%       ~ (p=0.656 n=30+20)
/dns-recordset/preview-initial/ResourceProvider/GetSchema/                               162.3m ±  1%    164.5m ±  1%  +1.36% (p=0.002 n=90+60)

```

I've organized the code into three parts:

1) traceutils.go contains functions and interfaces for working with traces and spans in their various formats. I expect we'll want to move most of this into a library, probably in pulumi-trace-tool (and add tests for it when we do)

2) benchmarks.go is the meat of the benchmark testing; it's patterned off of  the test config+runners in providertest, and can hopefully migrate there.

3) performance_test.go and provider_test.go are examples using the benchmark test on test programs. 

I've also checked in the recorded benchmarks from the tests in performance_test.go as a baseline and as examples of what the results look like.

Resolves https://github.com/pulumi/home/issues/3167

### CODE REVIEW

**Comment by mjeffryes** [comment]
FYI @VenelinMartinov This is what I ended up writing to collect benchmarks

File: provider/traceutils.go
Comment (VenelinMartinov) [inline provider/traceutils.go:106]: Does this ensure the struct conforms to the interface?

File: provider/benchmark.go
Comment (VenelinMartinov) [inline provider/benchmark.go:148]: 🥲

**Review by VenelinMartinov — APPROVED** [review]
🚀

File: provider/testdata/benchmarks/storage-bucket-bench.txt
Comment (t0yv0) [inline provider/testdata/benchmarks/storage-bucket-bench.txt:61]: Ah very interesting so the idea is to check in baselines? I think I like that. We have a similar problem with upgrade test recording baselines and the question of whether these belong in the repo still remains. We could consider alternate storage. Perhaps GitHub actions permits storing long-lived artifacts, or if that fails we can use an S3 bucket. But certainly not blocking on this, a great start.

File: provider/traceutils.go
Comment (t0yv0) [reply provider/traceutils.go:106]: It looks like that's the intent. The form I use these days is `var _ SpanIter = CsvFileSpanIter{}` is probably equivalent.

File: provider/benchmark.go
Comment (t0yv0) [inline provider/benchmark.go:121]: Fancy stuff from providertest? We're getting a bit carried away there. I'm open to ripping it out; in the end a simple side-by-side test for the benchmark is not that bad. A new top level `func TestXYZPerf()`.

**Review by t0yv0 — APPROVED** [review]
There's a fair bit of internals that can be tucked away not to be an eyesore in this repo, bu that can happen later. Should we check this in? Any reason not to? I feel there's no risk in having this.

File: provider/traceutils.go
Comment (mjeffryes) [reply provider/traceutils.go:106]: Yeah this trick threw me for a loop the first time I saw it too. Why have an explicit keyword for asserting compliance with an interface when you can just abuse casting apparently?

**Comment by mjeffryes** [comment]
Yea, I'm going to land this and we can work on factoring the mess out from GCP and making it reusable in the coming weeks. Since we're so close on the major version release, I might hold off until we land that branch, just to avoid adding merge conflicts for guin.

**Pull request merged**
